### PR TITLE
More CoreFx porting work

### PR DESCRIPTION
**Pull request by jaredpar**
Made a number of small changes to enable building on CoreFx

- Remove uses of MemoryStream::GetBuffer.  This API isn't supported on
CoreFx. The TryGetBuffer is available but in these cases seemed better
to just call Stream::CopyTo
- Changed Type.Assembly to Type.GetTypeInfo().Assembly
- Moved scripting APIs out of CommonCompiler and into
ScriptCompilerUtil.  These methods relied on APIs that are not available
on CoreFx and have no equivalent.
- Minor API updates.

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @gafter, @tmat, @agocke, @VSadov, @AlekseyTs

File: src/Compilers/Core/Desktop/CommandLine/CommonCompiler.cs [hunk 1]
```diff
@@ -230,7 +226,6 @@ protected bool ReportErrors(IEnumerable<Diagnostic> diagnostics, TextWriter cons
             return hasErrors;
         }
 
-        protected bool ReportErrors(IEnumerable<DiagnosticInfo> diagnostics, TextWriter consoleOutput, ErrorLogger errorLogger)
```
Comment (jaredpar) [inline src/Compilers/Core/Desktop/CommandLine/CommonCompiler.cs]: Strange this is being reported as a delete, I just changed the modifier to `public`

File: src/Compilers/Core/Desktop/CommandLine/CommonCompiler.cs [hunk 2]
```diff
@@ -261,7 +256,6 @@ protected virtual void PrintError(DiagnosticInfo diagnostic, TextWriter consoleO
             consoleOutput.WriteLine(diagnostic.ToString(Culture));
         }
 
-        private ErrorLogger GetErrorLogger(TextWriter consoleOutput, CancellationToken cancellationToken)
```
Comment (jaredpar) [inline src/Compilers/Core/Desktop/CommandLine/CommonCompiler.cs]: Same, just a modifier change to `public`

File: src/Compilers/Core/Desktop/CommandLine/RuleSet/RuleSet.cs [hunk 3]
```diff
@@ -209,7 +209,7 @@ private void GetEffectiveIncludesCore(ImmutableArray<string>.Builder arrayBuilde
                 }
 
                 // If this file has already been included don't recurse into it.
-                if (!arrayBuilder.Contains(ruleSet.FilePath, StringComparer.InvariantCultureIgnoreCase))
+                if (!arrayBuilder.Contains(ruleSet.FilePath, StringComparer.OrdinalIgnoreCase))
```
Comment (jaredpar) [inline src/Compilers/Core/Desktop/CommandLine/RuleSet/RuleSet.cs]: This does not exist in CoreFx and seemed incorrect anyways.  This is a file path comparison.

File: src/Scripting/Core/ScriptCompilerUtil.cs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Scripting
+{
+    internal static class ScriptCompilerUtil
```
Comment (jaredpar) [inline src/Scripting/Core/ScriptCompilerUtil.cs]: All of this code was moved from CommonCompiler.  There is no logic change.

File: src/Compilers/Core/Desktop/CommandLine/RuleSet/RuleSet.cs (hunk 3 above)
Comment (pharring) [inline src/Compilers/Core/Desktop/CommandLine/RuleSet/RuleSet.cs]: Agreed.

File: src/Compilers/Core/VBCSCompiler/BuildProtocol.cs [hunk 5]
```diff
@@ -184,7 +184,7 @@ internal class BuildRequest
                 // to prevent making another copy.
                 var stream = (MemoryStream)writer.BaseStream;
```
Comment (pharring) [inline src/Compilers/Core/VBCSCompiler/BuildProtocol.cs]: I think the cast to MemoryStream is now no longer necessary.

File: src/Compilers/Core/VBCSCompiler/BuildProtocol.cs [hunk 6]
```diff
@@ -259,8 +258,8 @@ public void WriteToBinaryWriter(BinaryWriter writer)
     ///
     /// Field Name       Field Type          Size (bytes)
     /// -------------------------------------------------
-    /// responseLength   int (positive)      4  
     /// responseType     enum ResponseType   4
+    /// responseLength   int (positive)      4
```
Comment (pharring) [inline src/Compilers/Core/VBCSCompiler/BuildProtocol.cs]: I don't think this change is correct. i.e. I think this was OK before.
The length is written first, then the "payload" from the BinaryWriter which leads with the 'type'

File: src/Compilers/Core/VBCSCompiler/BuildProtocol.cs [hunk 7]
```diff
@@ -292,8 +291,10 @@ public enum ResponseType
                 // Grab the MemoryStream and its internal buffer to prevent
                 // making another copy.
                 var stream = (MemoryStream)writer.BaseStream;
```
Comment (pharring) [inline src/Compilers/Core/VBCSCompiler/BuildProtocol.cs]: Looks like we no longer need to cast to MemoryStream

File: src/Scripting/Core/ScriptCompilerUtil.cs [hunk 8]
```diff
@@ -0,0 +1,118 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Emit;
```
Comment (pharring) [inline src/Scripting/Core/ScriptCompilerUtil.cs]: Sort usings?

File: src/Compilers/Core/VBCSCompiler/BuildProtocol.cs (hunk 6 above)
Comment (jaredpar) [inline src/Compilers/Core/VBCSCompiler/BuildProtocol.cs]: You're correct.  The use of two `Stream` types here was throwing me off.

File: src/Compilers/Core/VBCSCompiler/BuildProtocol.cs [hunk 9]
```diff
@@ -182,9 +182,9 @@ internal class BuildRequest
 
                 // Grab the MemoryStream and its internal buffer
                 // to prevent making another copy.
-                var stream = (MemoryStream)writer.BaseStream;
+                var stream = writer.BaseStream;
                 // Write the length of the request
-                int length = (int)stream.Length;
+                int length = checked((int)stream.Length);
```
Comment (gafter) [inline src/Compilers/Core/VBCSCompiler/BuildProtocol.cs]: This should not be checked. The very next line protects against the overflow case.

Comment (jaredpar) [inline src/Compilers/Core/VBCSCompiler/BuildProtocol.cs]: @gafter the next line checks for a maximum value, it does not check for overflow.

Comment (gafter) [inline src/Compilers/Core/VBCSCompiler/BuildProtocol.cs]: Ah, so the checked() is protecting against underflow (wrapping to a negative int).

**Comment by gafter** [comment]
LGTM

**Pull request closed without merging**